### PR TITLE
feat(speedup): add the concrete_trace_utils, to extend the torch.fx feature to trace  more models into fx.graph

### PR DESCRIPTION
**Pull request by Louis-J**
### Description ###

add a function concrete_trace to fx.
it is different from symbolic_trace that this tracer concrete execute all the python instructions, so we have real value such as dim, shape and then we can pass the branches like `if x.dim() == 4:`.

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [x] test case
  - [ ] doc

### CODE REVIEW

File: nni/common/concrete_trace_utils/concrete_proxy.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
+from __future__ import annotations
```
Comment (liuzhe-lz) [inline nni/common/concrete_trace_utils/concrete_proxy.py:1]: Missing copyright header.

File: nni/common/concrete_trace_utils/concrete_proxy.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            # This allows you to create a ConcreteProxy object around a raw Node
+            tracer = et.GraphAppendingConcreteTracer(node.graph)
+        self.tracer = tracer
+        self.value = value
+        self.node = node
+
+    def __repr__(self) -> str:
+        return f'ConcreteProxy({self.node.name})'
+
+    def __getattr__(self, k) -> 'ConcreteProxy':
```
Comment (liuzhe-lz) [inline nni/common/concrete_trace_utils/concrete_proxy.py:44]: With `__future__.annotations` there should be no quotes in type hint.

File: nni/common/concrete_trace_utils/concrete_proxy.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        assert frame is not None
+        calling_frame = frame.f_back
+        assert calling_frame is not None
+        insts = list(dis.get_instructions(calling_frame.f_code))
+        cur = calling_frame.f_lasti // 2
+        while insts[cur].opcode == dis.opmap['EXTENDED_ARG']:
+            cur += 1
+
+        if insts[cur].opcode in self.jump_opcodes:
+            return self.value.__bool__()
```
Comment (liuzhe-lz) [inline nni/common/concrete_trace_utils/concrete_proxy.py:100]: If there is no difference, `bool(self.value)` should be preferred.

File: nni/common/concrete_trace_utils/concrete_proxy.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        elif insts[cur].opcode == dis.opmap['UNARY_NOT']:
+            # log warning
+            print('please use the function patcher, or use "from operator import not_; x = not_(y)" instead of "x = not y", or the traced graph may be wrong')
```
Comment (liuzhe-lz) [inline nni/common/concrete_trace_utils/concrete_proxy.py:103]: Use `logger.warning()` if it works.

File: nni/common/concrete_trace_utils/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
```
Comment (matluster) [inline nni/common/concrete_trace_utils/__init__.py:3]: Suggest adding module-level docstrings to explain what this is all about.

File: examples/concretc_tracer/concrete_tracer_mmdetection.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (Louis-J) [reply examples/concretc_tracer/concrete_tracer_mmdetection.py:1]: concrete

File: examples/concretc_tracer/concrete_tracer_mmdetection.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+from mmdet.apis import init_detector
+from mmcv.parallel import collate
+from mmdet.datasets import replace_ImageToTensor
+from mmdet.datasets.pipelines import Compose
+from nni.common.concrete_trace_utils import concrete_trace, ConcreteTracer
+
+folder_prefix = 'C:\\Works\\mmdetection' # replace this path with yours
+img = '%s/tests/data/color.jpg' % folder_prefix
+
+config_files_correct = (
```
[the remark this answers is not in the document]
Comment (Louis-J) [reply examples/concretc_tracer/concrete_tracer_mmdetection.py:18]: ok

File: nni/common/concrete_trace_utils/__init__.py (hunk 5 above)
Comment (Louis-J) [reply nni/common/concrete_trace_utils/__init__.py:3]: ok

File: examples/concretc_tracer/concrete_tracer_mmdetection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matluster) [reply examples/concretc_tracer/concrete_tracer_mmdetection.py:18]: Depends on how long it runs.

I suppose it won't hurt to put it into "full test - compression".

File: pipelines/full-test-compression.yml [hunk 8]
```diff
@@ -42,5 +42,15 @@ stages:
 
     - script: |
         cd test/algo
+        pwd
+        rm -rf ../mmdetection
+        git clone https://github.com/open-mmlab/mmdetection.git ../mmdetection
```
Comment (matluster) [inline pipelines/full-test-compression.yml:47]: You might use [Azure pipeline repo checkout syntax](https://learn.microsoft.com/en-us/azure/devops/pipelines/repos/multi-repo-checkout?view=azure-devops) to checkout a specific version of mmdetection. Better to avoid using master here because master can be potentially unstable.

File: test/algo/compression/pruning/test_concrete_trace_mmdetection.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+            mmcv_cnn.bricks.wrappers.MaxPool3d,
+        ), fake_middle_class = (
+            mmdet_core.anchor.anchor_generator.AnchorGenerator,
+        ))
+
+        if config_file == 'pvt/retinanet_pvt-l_fpn_1x_coco':
+            torch_fx.proxy.base_types = orig_base_types
+
+        seed = torch.seed()
+        torch.manual_seed(seed)
```
Comment (matluster) [inline test/algo/compression/pruning/test_concrete_trace_mmdetection.py:297]: If seed is manually set, it should be restored to a random seed (e.g., based on timestamp) after this test so that other tests won't be affected.

File: examples/concretc_tracer/concrete_trace_distilbert.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import torch
```
Comment (matluster) [inline examples/concretc_tracer/concrete_trace_distilbert.py:1]: The folder name (examples/concretc_tracer) is still misspelled.

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 11]
```diff
@@ -0,0 +1,1256 @@
...
+            else:
+                try:
+                    return _orig_clz_getattribute(obj, attr)
+                except AttributeError:
+                    return _orig_clz_getattr(obj, attr)
+        else:
+            return tracer.create_proxy('get_attr', f'{the_path_of_middle_class[id(obj)]}.{attr}', (), {})
+    return clz_getattr_wrapper
+
+def concrete_trace(root : Union[torch.nn.Module, Callable[..., Any]],
```
Comment (matluster) [inline nni/common/concrete_trace_utils/concrete_tracer.py:1212]: This is the top-level API, right?

I think it deserves some docstrings to explain:

- Who should use this?
- What are its expected input arguments?
- What is its return type? How to interpret / use its return value?

File: nni/common/concrete_trace_utils/__init__.py [hunk 12]
```diff
@@ -0,0 +1,4 @@
...
+from .concrete_tracer import ConcreteTracer, concrete_trace
```
Comment (matluster) [inline nni/common/concrete_trace_utils/__init__.py:4]: Please add the APIs to API reference page in docs.

File: test/algo/compression/pruning/test_concrete_trace_mmdetection.py (hunk 9 above)
Comment (Louis-J) [reply test/algo/compression/pruning/test_concrete_trace_mmdetection.py:297]: The seed comes from `torch.seed` which would return a random seed, So subsequent tests will not be affected.

File: nni/common/concrete_trace_utils/operator_patcher.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+            return self.generic_visit(node)
+        should_replace = False
+        for op in node.ops:
+            if type(op) in (ast.Is, ast.IsNot, ast.In, ast.NotIn):
+                should_replace = True
+                break
+        if should_replace:
+            if len(node.ops) != 1:
+                raise RuntimeError(
+                    'not supported in "{} cmp_op {} cmp_op {}" when cmp_op contains "is/is not/in/not in"')
```
[the remark this answers is not in the document]
Comment (Louis-J) [reply nni/common/concrete_trace_utils/operator_patcher.py:122]: some strange but supported syntax such as:
```python
  a = 1
  b = (1,2)
  c = (b, (3,4))
  d = a in b in c # strange but legal
  # equlas to: d = a in b and b in c
```

File: test/algo/compression/pruning/test_concrete_trace_mmdetection.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+import mmcv
+import mmcv.cnn as mmcv_cnn
+import mmdet.core as mmdet_core
+from mmdet.apis import init_detector
+from mmcv.parallel import collate
+from mmdet.datasets import replace_ImageToTensor
+from mmdet.datasets.pipelines import Compose
+from nni.common.concrete_trace_utils import concrete_trace, ConcreteTracer
+
+assert 'MMDET_DIR' in os.environ, 'please set env variable `MMDET_DIR` to your mmdetection folder!'
```
Comment (matluster) [inline test/algo/compression/pruning/test_concrete_trace_mmdetection.py:18]: Suggest `pytest.skip` if MMDET is not found. So that the test collection doesn't fail simply because they didn't set MMDET_DIR.

File: test/algo/compression/pruning/test_concrete_trace_mmdetection.py (hunk 9 above)
Comment (matluster) [reply test/algo/compression/pruning/test_concrete_trace_mmdetection.py:269]: If seed is a random seed, why bother? You could simply omit setting seed.

**Review by ultmaster — APPROVED** [review]

**Review by J-shang — APPROVED** [review]
